File: translations_only/fail/go/triangolatte/Line/gemini/Restart/Line.rs

use std::f64::consts::PI;
use std::f64::EPSILON;
use std::ops::{Add, Sub, Mul, Neg};

#[derive(Copy, Clone, Debug)]
struct Point {
    x: f64,
    y: f64,
}

impl Point {
    fn new(x: f64, y: f64) -> Self {
        Point { x, y }
    }

    fn scale(&self, f: f64) -> Self {
        let norm = (self.x * self.x + self.y * self.y).sqrt();
        Point {
            x: self.x / norm * f,
            y: self.y / norm * f,
        }
    }

    fn add(&self, r: &Point) -> Self {
        Point {
            x: self.x + r.x,
            y: self.y + r.y,
        }
    }

    fn sub(&self, r: &Point) -> Self {
        Point {
            x: self.x - r.x,
            y: self.y - r.y,
        }
    }

    fn dot(&self, r: &Point) -> f64 {
        self.x * r.x + self.y * r.y
    }

    fn normalize(&self) -> Self {
        self.scale(1.0)
    }
}

#[derive(Copy, Clone, Debug)]
enum Joint {
    Normal,
    Miter,
}

fn line(joint: Joint, points: &[Point], width: f64) -> Vec<f64> {
    match joint {
        Joint::Normal => normal(points, width),
        Joint::Miter => miter(points, width),
    }
}

fn normal(points: &[Point], width: f64) -> Vec<f64> {
    let width = width / 2.0;
    let mut triangles = Vec::with_capacity(points.len() * 12);
    for i in 0..points.len() - 1 {
        let dx = points[i + 1].x - points[i].x;
        let dy = points[i + 1].y - points[i].y;
        let n1 = Point::new(dy, -dx).scale(width);
        let n2 = Point::new(-dy, dx).scale(width);

        let v0 = points[i + 1].add(&n2).x;
        let v1 = points[i + 1].add(&n2).y;
        let v2 = points[i].add(&n2).x;
        let v3 = points[i].add(&n2).y;
        let v4 = points[i].add(&n1).x;
        let v5 = points[i].add(&n1).y;
        let v6 = points[i].add(&n1).x;
        let v7 = points[i].add(&n1).y;
        let v8 = points[i + 1].add(&n1).x;
        let v9 = points[i + 1].add(&n1).y;
        let v10 = points[i + 1].add(&n2).x;
        let v11 = points[i + 1].add(&n2).y;

        triangles.extend_from_slice(&[v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11]);
    }

    triangles
}

fn miter(points: &[Point], width: f64) -> Vec<f64> {
    let width = width / 2.0;
    let mut triangles = Vec::with_capacity(points.len() * 12);
    let mut dx: f64;
    let mut dy: f64;
    let mut miter1 = Point::new(0.0, 0.0);
    let mut miter2 = Point::new(0.0, 0.0);
    let mut n1: [Point; 2];
    let mut n2: [Point; 2];

    dx = points[1].x - points[0].x;
    dy = points[1].y - points[0].y;

    n2 = calculate_normals(dx, dy);
    miter2 = n2[0].scale(width);

    for i in 1..points.len() - 1 {
        // Shift calculated values.
        n1 = n2;
        miter1 = miter2;

        dx = points[i + 1].x - points[i].x;
        dy = points[i + 1].y - points[i].y;

        n2 = calculate_normals(dx, dy);

        // Find tangent vector to both lines in the middle point.
        let tangent = (points[i + 1].sub(&points[i])).normalize().add(&(points[i].sub(&points[i - 1])).normalize()).normalize();

        // Miter vector is perpendicular to the tangent and crosses extensions of
        // normal-translated lines in miter join points.
        let unit_miter = Point::new(-tangent.y, tangent.x);

        // Length of the miter vector projected onto one of the normals.
        // Choice of normal is arbitrary, each of them would work.
        let miter_length = width / unit_miter.dot(&n1[0]);
        miter2 = unit_miter.scale(miter_length);

        let v0 = points[i].sub(&miter2).x;
        let v1 = points[i].sub(&miter2).y;
        let v2 = points[i - 1].sub(&miter1).x;
        let v3 = points[i - 1].sub(&miter1).y;
        let v4 = points[i - 1].add(&miter1).x;
        let v5 = points[i - 1].add(&miter1).y;
        let v6 = points[i - 1].add(&miter1).x;
        let v7 = points[i - 1].add(&miter1).y;
        let v8 = points[i].add(&miter2).x;
        let v9 = points[i].add(&miter2).y;
        let v10 = points[i].sub(&miter2).x;
        let v11 = points[i].sub(&miter2).y;

        triangles.extend_from_slice(&[v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11]);
    }

    // Use last normal as another 'neutral element' for miter join.
    let n = points.len();
    let last_miter = n2[0].scale(width);

    let v0 = points[n - 1].sub(&last_miter).x;
    let v1 = points[n - 1].sub(&last_miter).y;
    let v2 = points[n - 2].sub(&miter1).x;
    let v3 = points[n - 2].sub(&miter1).y;
    let v4 = points[n - 2].add(&miter1).x;
    let v5 = points[n - 2].add(&miter1).y;
    let v6 = points[n - 2].add(&miter1).x;
    let v7 = points[n - 2].add(&miter1).y;
    let v8 = points[n - 1].add(&last_miter).x;
    let v9 = points[n - 1].add(&last_miter).y;
    let v10 = points[n - 1].sub(&last_miter).x;
    let v11 = points[n - 1].sub(&last_miter).y;

    triangles.extend_from_slice(&[v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11]);

    triangles
}

fn calculate_normals(x: f64, y: f64) -> [Point; 2] {
    [
        Point::new(y, -x).normalize(),
        Point::new(-y, x).normalize(),
    ]
}

File: bms/go/triangolatte/Line/Line.go
package test

import (
	"errors"
	"math"
)

var ()

const (
	// Normal triangulates with no joint correction.
	Normal Joint = 0
	// Miter triangulates producing miter joints, i.e. extending the lines until
	// they meet at some point.
	Miter Joint = 1
)

func Line(joint Joint, points []Point, width float64) ([]float64, error) {
	switch joint {
	case Normal:
		return normal(points, width), nil
	case Miter:
		return miter(points, width), nil
	default:
		return nil, errors.New("Unrecognized joint type")
	}
}

func normal(points []Point, width float64) []float64 {
	width /= 2.0
	triangles := make([]float64, 0, len(points)*12)
	for i := 0; i <= len(points)-2; i++ {
		dx := points[i+1].X - points[i].X
		dy := points[i+1].Y - points[i].Y
		n1 := Point{dy, -dx}.Scale(width)
		n2 := Point{-dy, dx}.Scale(width)

		v0, v1 := points[i+1].Add(n2).X, points[i+1].Add(n2).Y
		v2, v3 := points[i].Add(n2).X, points[i].Add(n2).Y
		v4, v5 := points[i].Add(n1).X, points[i].Add(n1).Y
		v6, v7 := points[i].Add(n1).X, points[i].Add(n1).Y
		v8, v9 := points[i+1].Add(n1).X, points[i+1].Add(n1).Y
		v10, v11 := points[i+1].Add(n2).X, points[i+1].Add(n2).Y

		triangles = append(triangles, v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11)
	}

	return triangles
}

func (p Point) Scale(f float64) Point {
	norm := float64(math.Sqrt(float64(p.X*p.X + p.Y*p.Y)))
	return Point{p.X / norm * f, p.Y / norm * f}
}

func (p Point) Add(r Point) Point {
	return Point{p.X + r.X, p.Y + r.Y}
}

func miter(points []Point, width float64) []float64 {
	width /= 2.0
	triangles := make([]float64, 0, len(points)*12)
	var dx, dy float64
	var miter1, miter2 Point
	var n1, n2 [2]Point

	dx = points[1].X - points[0].X
	dy = points[1].Y - points[0].Y

	n2 = calculateNormals(dx, dy)
	miter2 = n2[0].Scale(width)

	for i := 1; i < len(points)-1; i++ {
		// Shift calculated values.
		n1 = n2
		miter1 = miter2

		dx = points[i+1].X - points[i].X
		dy = points[i+1].Y - points[i].Y

		n2 = calculateNormals(dx, dy)

		// Find tangent vector to both lines in the middle point.
		tangent := (points[i+1].Sub(points[i])).Normalize().Add((points[i].Sub(points[i-1])).Normalize()).Normalize()

		// Miter vector is perpendicular to the tangent and crosses extensions of
		// normal-translated lines in miter join points.
		unitMiter := Point{-tangent.Y, tangent.X}

		// Length of the miter vector projected onto one of the normals.
		// Choice of normal is arbitrary, each of them would work.
		miterLength := float64(width) / unitMiter.Dot(n1[0])
		miter2 = unitMiter.Scale(miterLength)

		v0, v1 := points[i].Sub(miter2).X, points[i].Sub(miter2).Y
		v2, v3 := points[i-1].Sub(miter1).X, points[i-1].Sub(miter1).Y
		v4, v5 := points[i-1].Add(miter1).X, points[i-1].Add(miter1).Y
		v6, v7 := points[i-1].Add(miter1).X, points[i-1].Add(miter1).Y
		v8, v9 := points[i].Add(miter2).X, points[i].Add(miter2).Y
		v10, v11 := points[i].Sub(miter2).X, points[i].Sub(miter2).Y

		triangles = append(triangles, v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11)
	}

	// Use last normal as another 'neutral element' for miter join.
	n := len(points)
	lastMiter := n2[0].Scale(width)

	v0, v1 := points[n-1].Sub(lastMiter).X, points[n-1].Sub(lastMiter).Y
	v2, v3 := points[n-2].Sub(miter1).X, points[n-2].Sub(miter1).Y
	v4, v5 := points[n-2].Add(miter1).X, points[n-2].Add(miter1).Y
	v6, v7 := points[n-2].Add(miter1).X, points[n-2].Add(miter1).Y
	v8, v9 := points[n-1].Add(lastMiter).X, points[n-1].Add(lastMiter).Y
	v10, v11 := points[n-1].Sub(lastMiter).X, points[n-1].Sub(lastMiter).Y

	triangles = append(triangles, v0, v1, v2, v3, v4, v5, v6, v7, v8, v9, v10, v11)

	return triangles
}

func calculateNormals(x, y float64) [2]Point {
	return [2]Point{
		Point{y, -x}.Normalize(),
		Point{-y, x}.Normalize(),
	}
}

func (p Point) Normalize() Point {
	return p.Scale(1)
}

func (p Point) Sub(r Point) Point {
	return Point{p.X - r.X, p.Y - r.Y}
}

func (p Point) Dot(r Point) float64 {
	return p.X*r.X + p.Y*r.Y
}

type Joint int

type Point struct {
	X, Y float64
}
